libs/utils/android/benchmark.py

```diff
@@ -0,0 +1,262 @@
+        Set up logging and trigger running experiments
+        """
+        LisaLogging.setup()
+        self.log = logging.getLogger('Benchmark')
+
+        self.log.info('=== CommandLine parsing...')
+        self._parseCommandLine()
+
+        self.log.info('=== TestEnv setup...')
+        self.te = TestEnv(self._getBmConf())
+        self.target = self.te.target
+        # TODO: Check if TestEnv already provides support for rebooting
+        self._reboot()
+
+        self.log.info('=== Initialization...')
+        self._getBmKind()
+        self.out_dir=self.te.res_dir
+        if not self.benchmarkInit():
+            self.log.debug('Initialization failed: execution aborted')
+            return
+
+        self.log.info('=== Execution...')
+        self.wl.run(out_dir=self.out_dir,
+                    collect=self._getBmCollect(),
+                    **self.bm_params)
+
+        self.log.info('=== Finalization...')
+        self.benchmarkFinalize()
+
+    def _adb(self, cmd):
+        os.system('adb -s {} {}'.format(self.target.adb_name, cmd))
+
+    def _fastboot(self, cmd):
+        os.system('fastboot -s {} {}'.format(self.target.adb_name, cmd))
+
+
+    def _wait_for_logcat_idle(self, seconds=1):
+        lines = 0
+
+        # Clear logcat
+        # os.system('{} logcat -s {} -c'.format(adb, DEVICE));
+        self._adb('logcat -c')
+
+        # Dump logcat output
+        logcat_cmd = 'adb -s {} logcat'.format(self.target.adb_name)
+        logcat = Popen(logcat_cmd, shell=True, stdout=PIPE)
+        logcat_poll = select.poll()
+        logcat_poll.register(logcat.stdout, select.POLLIN)
+
+        # Monitor logcat until it's idle for the specified number of [s]
+        self.log.info('Waiting for system to be almost idle')
+        self.log.info('   i.e. at least %d[s] of no logcat messages', seconds)
+        while True:
+            poll_result = logcat_poll.poll(seconds * 1000)
+            if not poll_result:
+                break
+            lines = lines + 1
+            line = logcat.stdout.readline(1024)
+
+    def _reboot(self):
+
+        # Reboot the device, if a boot_image has been specified
+        if self.args.boot_image:
+
+            self.log.warning('=== Rebooting...')
+            self.log.warning('Rebooting image to use: %s', self.args.boot_image)
+
+            self.log.debug('Waiting 6[s] to enter bootloader...')
+            self._adb('reboot-bootloader')
+            sleep(6)
+            self._fastboot('boot {}'.format(self.args.boot_image))
+
+            self.log.debug('Waiting 20[s] for boot to start...')
+            sleep(20)
+
+        else:
+            self.log.warning('Device NOT rebooted, using current image')
+
+        # Restart ADB in root mode
+        self._adb('root')
+
+        # TODO add check for kernel SHA1
+
+        # Disable charge via USB
+        self.log.debug('Disabling charge over USB...')
+        self._adb('shell "echo 0 >/sys/class/power_supply/battery/charging_enabled"')
```

Comment:
> Not really, I think we still need it (at least for this benachmark) to be able collecting db files generated in the app folder... moreover, in any case, I would say that devices we want to use for tests have to be rooted as a general rule.

OK.

>What about adding this directly as a pretty much standard target's API? 

Funnily enough I wrote patches for devlib for this at some point https://github.com/bjackman/devlib/commit/96d0e610441410d40f223f31836f24df4e8ff74d but haven't got round to submitting them yet. Not sure how generic it is.